### ISSUE TITLE
Very low mAP

### ISSUE THREAD

**Issue by pandasMX** [AlexeyAB/darknet#1749]
Hi, I trained a yolov2 with a 3 classes dataset. The loss went down to around 0.3 after 30k steps. However, when i ran the `detector valid`, the output in results folder seems to be wrong. 
I got comp4_det_test_0.txt, comp4_det_test_1.txt, comp4_det_test_2.txt. But two of them were empty. And I think this resulted in the very low mAP. I only got 0.9% mAP.

My config files are here:

3cat.data
`classes = 3
train  = /data/YOLO-3cat/train.txt
valid  = /data/YOLO-3cat/test.txt
names = data/3cat.names
backup = backup`

3cat.names
```
0
1
2
```

3cat.cfg
```
[net]
# Testing
batch=32
subdivisions=16
# Training
#batch=32
#subdivisions=16
height=416
width=416
channels=3
momentum=0.9
decay=0.0005
angle=0
saturation = 1.5
exposure = 1.5
hue=.1

learning_rate=0.001
burn_in=1000
max_batches = 180200
policy=steps
steps=15000,30000,60000,100000
scales=.5,.5,.1,.1

...

[convolutional]
size=1
stride=1
pad=1
filters=40
activation=linear


[region]
anchors =  1.3221, 1.73145, 3.19275, 4.00944, 5.05587, 8.09892, 9.47112, 4.84053, 11.2364, 10.0071
bias_match=1
classes=3
coords=4
num=5
softmax=1
jitter=.3
rescore=1

object_scale=5
noobject_scale=1
class_scale=1
coord_scale=1

absolute=1
thresh = .6
random=1
```

samples in train.txt and test.txt:
```
/data/YOLO-3cat/JPEGImages/c3c2c109f974dcd39ca704529613e977.jpg
/data/YOLO-3cat/JPEGImages/232f44d5107e6b1a8f98707688c3f734.jpg
/data/YOLO-3cat/JPEGImages/ffe0452c9a12ebb9aac1689977ea5239.jpg
```

sample .txt file:
`0 0.50341796875 0.49072265625 0.9365234375 0.6708984375`

**Comment by AlexeyAB** [comment]
@pandasMX Hi,

>  I only got 0.9% mAP.

* Can you show screenshot of the command `./darknet detector map obj.data `.... ?

* Can you show content of the file `obj.data`?

**Comment by pandasMX** [comment]
Hi, I solved the problem by changing CUDNN=0.
During training I used CUDNN=1 in Makefile, but this will make detector valid generating random results. After I change to CUDNN=0 and make, then run detector valid, I got 80%mAP. 

So my thought is to set CUDNN=1 during training to speed up, then change to CUDNN=0 when you want to evaluate your mAP.

**Comment by AlexeyAB** [comment]
This is very strange.
CUDNN=1 should have the same result as CUDNN=0.

* What GPU, CUDA, cuDNN do you use?

* Can you show output of this command: `nvidia-smi` ?

**Comment by AlexeyAB** [comment]
* Also show output of `nvcc --version` and check that you use cuDNN version that is appropriate for your CUDA version: https://developer.nvidia.com/rdp/cudnn-archive

* You can try to use this repository https://github.com/AlexeyAB/darknet with flag `GPU=1 CUDNN=1 CUDNN_HALF=1 OPENCV=1` it should work 2x-3x times faster than original repository https://github.com/pjreddie/darknet

* If it doesn't help - try to install new nVidia GPU driver: https://www.nvidia.ru/Download/index.aspx?lang=en

**Comment by leo-XUKANG** [comment]
@AlexeyAB  I GOT the same problem
when i set the parameters
GPU=1
CUDNN=0
OPENCV=1
OPENMP=1
DEBUG=0

i got

for thresh = 0.50, precision = 0.78, recall = 0.63, F1-score = 0.70
for thresh = 0.50, TP = 7569, FP = 2075, FN = 4463, average IoU = 57.87 %

mean average precision (mAP) = 0.812613, or 81.26 %

but when
GPU=1
CUDNN=1
OPENCV=1
OPENMP=1
DEBUG=0

for thresh = 0.50, precision = 0.77, recall = 0.00, F1-score = 0.01
for thresh = 0.50, TP = 36, FP = 11, FN = 11996, average IoU = 50.37 %

mean average precision (mAP) = 0.064466, or 6.45 %

**Comment by AlexeyAB** [comment]
@leo-XUKANG 

* What repository do you use https://github.com/AlexeyAB/darknet ro https://github.com/pjreddie/darknet ?

* Show full command that you use for getting mAP.

* What CUDA, cuDNN, OpenCV versions do you use?

* Show output of `nvidia-smi` command.

**Comment by leo-XUKANG** [comment]
1.   I use the repository  https://github.com/pjreddie/darknet,  but after i read your work , I make it possible to use the command " ./darknet detector map  voc_data_path  cfg_file_path  weights_file_path" 

I changed the detector.c file according to your work,  i  have to say that your work is so beautiful

2.    the command is   ./darknet detector map  voc_data_path  cfg_file_path  weights_file_path 

3.  cuda8.0  cudnn7.0  opencv3.1.0

4.  +-----------------------------------------------------------------------------+
| NVIDIA-SMI 390.46                 Driver Version: 390.46                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla P100-PCIE...  On   | 00000000:02:00.0 Off |                    0 |
| N/A   56C    P0    42W / 250W |   9325MiB / 12198MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   1  Tesla P100-PCIE...  On   | 00000000:03:00.0 Off |                    0 |
| N/A   58C    P0    42W / 250W |   9325MiB / 12198MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   2  Tesla P100-PCIE...  On   | 00000000:83:00.0 Off |                    0 |
| N/A   61C    P0    41W / 250W |   9906MiB / 12198MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   3  Tesla P100-PCIE...  On   | 00000000:84:00.0 Off |                    0 |
| N/A   58C    P0    38W / 250W |   9325MiB / 12198MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0      1372      C   ./darknet                                   9307MiB |
|    1      1372      C   ./darknet                                   9307MiB |
|    2      1372      C   ./darknet                                   9307MiB |
|    2     31842      C   /cm/shared/apps/matlab/bin/glnxa64/MATLAB    581MiB |
|    3      1372      C   ./darknet                                   9307MiB |
+-----------------------------------------------------------------------------+

**Comment by AlexeyAB** [comment]
@leo-XUKANG What mAP can you get by using my repository https://github.com/AlexeyAB/darknet ?

**Comment by leo-XUKANG** [comment]
@AlexeyAB 
you add a function in detector.c file, the function which will return the map,TP,FP,IOU  and so on.
https://github.com/AlexeyAB/darknet/blob/14ed6fcb6e31dd111fc5c35c31ffa6e45fe52737/src/detector.c#L559

i change the function , 
it could work in the repostitory https://github.com/pjreddie/darknet
and i test the original yolov2-voc, yolo-voc, 
it is ok

**Comment by leo-XUKANG** [comment]
@AlexeyAB 
i get the reason, the cuda version does not match the cudnn's
